llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1701,6 +1701,31 @@ static Instruction *foldFDivPowDivisor(BinaryOperator &I,
   return BinaryOperator::CreateFMulFMF(Op0, Pow, &I);
 }
 
+/// Convert div to mul if we have an sqrt divisor iff sqrt's operand is a fdiv
+/// instruction.
+static Instruction *foldFDivSqrtDivisor(BinaryOperator &I,
+                                        InstCombiner::BuilderTy &Builder) {
+  // X / sqrt(Y / Z) -->  X * sqrt(Z / Y)
+  Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
+  auto *II = dyn_cast<IntrinsicInst>(Op1);
+  if (!II || II->getIntrinsicID() != Intrinsic::sqrt || !II->hasOneUse() ||
+      !I.hasAllowReassoc() || !I.hasAllowReciprocal())
```

Comment:
Do we need `reassoc`?

Reply:
my assumption was following  the implementation of `foldFDivPowDivisor` or at least the div inside the sqrt should have that flag enabled?

Reply:
Using `x.recip()` as shorthand for `(1.0 / x)`, `arcp` allows us to convert between `a / b` and `a * b.recip()` freely, and arguably `(a / b).recip()` to `(b / a)` (at the very least, gcc will do this with just `-freciprocal-math`).

We can convert `x / sqrt(y / z)` to `x * sqrt(y / z).recip()` with just `arcp`, and we can also convert `x * sqrt((y / z).recip())` to `x * sqrt(z / y)` with just `arcp` as well, but the question is which flags are necessary to convert `sqrt(a).recip()` to `sqrt(a.recip())`. I think it stretches `arcp` too far to allow this kind of permutation of `recip`. `reassoc` isn't entirely the right flag either; what we probably want is a more generic "allow algebraic identity" flag, but we're already using `reassoc` for that purpose elsewhere, so we might as well use it here as well.

Reply:
okay so from my understanding then the pattern we are looking for is the following 

```
%div = fdiv arcp double %y, %z
%sqrt = call reassoc double @llvm.sqrt.f64(double %div)
%div2 = fdiv arcp double %x, %sqrt
```

Reply:
Not quite, you'd need the `reassoc` on the first `fdiv` as well, and at that point, it's probably best to just require that all the operations have `reassoc` and `arcp`.